libraries/drivers_appserver_base.rb

```diff
@@ -37,18 +40,30 @@ def appserver_config
         raise NotImplementedError
       end
 
+      # Creates a monit config file for managing the appserver and then notifies
+      # monit to reload it.
       def add_appserver_monit
-        opts = { app_shortname: app['shortname'], appserver_name: adapter, appserver_command: appserver_command,
-                 deploy_to: deploy_dir(app), environment: environment }
-
-        file_path = File.join(node['monit']['basedir'], "#{opts[:appserver_name]}_#{opts[:app_shortname]}.monitrc")
+        opts = {
+          app_shortname: app['shortname'],
+          adapter: adapter,
+          appserver_command: appserver_command,
+          appserver_name: adapter,
+          deploy_to: deploy_dir(app),
+          environment: environment,
+          source_cookbook: appserver_monit_template_cookbook
+        }
+        file_path = File.join(node['monit']['basedir'],
+                              "#{opts[:appserver_name]}_#{opts[:app_shortname]}.monitrc")
         context.template file_path do
           mode '0640'
-          source 'appserver.monitrc.erb'
+          source "#{opts[:adapter]}.monitrc.erb"
+          cookbook opts[:source_cookbook].to_s
           variables opts
+          notifies :run, 'execute[monit reload]', :immediately
```

Comment:
between this and setup there's a gotcha that's burned me in production twice now, because a Rails app server should not be started until all deploy tasks are completed; most obviously, asset compilation, but more besides (e.g. I have bespoke secrets injection at deploy time).

Reply:
I think the way to finesse this is with the `onreboot` monit option.

If we mark the service in monit as `onreboot nostart` until all deploy tasks are complete, then the service can be programatically stopped & started when we want, but monit won't auto-start it prematurely.

Once deploy tasks are completed and the service is ready, we'd set it back to the default of `onreboot start`, so that it still comes up on soft reboot.

Reply:
@inopinatus Interesting approach.  I'll see if I can get that to work.

What I see right now in my testing is that there are times when monit will attempt to start the rails server before it is properly configured (and the code may not be there either), but it just fails. Once the deploy happens and we trigger the `start` or `restart` via monit everything should be in place and work fine.

Is there something custom about your setup that would prevent this approach from working?

Reply:
@olbrich nothing custom. The problem is that monit can attempt to start the server after the code is in place, but before asset compilation has run/completed. Rails will start, but error out as soon as it attempts to consult the asset pipeline.